### PR TITLE
BFB-551: Create system for updating tunable values per Derek phase

### PR DESCRIPTION
**Pull request by Gursaanj**
**JIRA Ticket**
https://perigongames.atlassian.net/browse/BFB-551

**Description**
Created a system so that we can change the challenge/tunable values of Derek per their phase (Phase 1, 2 and 3). This system requires a SO data asset for tuning variables

Things we can change per phase with this PR:

- Health threshold of the phase (the health percentage at which the phase ends)
- Rotation rate of Derek
- Time Derek is vulnerable for when shields are down
- Interval length for consecutive missiles
- A speed modifier for the missiles launched (Only affects missiles after they reach apex and are homing towards the player)
- Ring Labor configurations (so we can set different puzzles per phase)

With different ring systems being needed to be setup based on which phase we are on, we have to initialize the system multiple times. So I moved the responsibility of Initializing and resetting the RingLaborManager to the ContextManager, instead of the WorldManager. We have this backlog ticket to address a more generic way of calling labor manager 

The missile rotation does look a bit weird at high multiplier values, but this will be addressed in a future bug fix!

**Screenshots / GIFs**
Configurations

We can make multiple Ring Systems (not the prettiest editor to look at haha) to handle multiple configurations

Phase SO

The base SO for editing tunable data per phase. Will be updated as we do future tuning passes

### CODE REVIEW

File: Assets/_BForBoss/_Weapons/Scripts/Bullets/BulletBehaviour.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ public abstract partial class BulletBehaviour : MonoBehaviour, IBullet
         protected bool _isActive = false;
 
         public Transform HomingTarget { get; set; }
+        public float SpeedMultiplier { get; set; } = 1.0f;
```
Comment (m4rtint) [inline Assets/_BForBoss/_Weapons/Scripts/Bullets/BulletBehaviour.cs:18]: Where is this used?

File: Assets/_BForBoss/_Weapons/Scripts/Bullets/IBullet.cs [hunk 2]
```diff
@@ -7,6 +7,7 @@ public interface IBullet
     {
         void SetSpawnAndDirection(Vector3 location, Vector3 normalizedDirection);
         Transform HomingTarget { set; }
+        float SpeedMultiplier { set; }
```
Comment (m4rtint) [inline Assets/_BForBoss/_Weapons/Scripts/Bullets/IBullet.cs:10]: Same here, where is this used?

File: Assets/_BForBoss/_Core/Scripts/Boss/DerekMissileLauncherBehaviour.cs [hunk 3]
```diff
@@ -33,10 +35,23 @@ public void Initialize(IGetPlayerTransform playerTransform)
                 _launcherVFX[i] = _launcherLocations[i].gameObject.GetComponentInChildren<TimedVFXEffect>();
             }
         }
-        
+
+        public void UpdateMissileSettings(float missileSpeedMultiplier, float intervalBetweenShots)
+        {
+            if (intervalBetweenShots <= 0.0f || missileSpeedMultiplier <= 0.0f)
```
[the remark this answers is not in the document]
Comment (Gursaanj) [reply Assets/_BForBoss/_Core/Scripts/Boss/DerekMissileLauncherBehaviour.cs:41]: Good point, nothing happens if it is zero I guess, the shots become very glitchy, Already added a range to them in the SO, so I can get rid of this check

File: Assets/_BForBoss/_Core/Scripts/Boss/DerekPhaseDataSO.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+using UnityEngine;
+
+namespace BForBoss
+{
+    [CreateAssetMenu(fileName = "Phase Data", menuName = "PerigonGames/Boss/Derek/PhaseData", order = 1)]
+    public class DerekPhaseDataSO : SerializedScriptableObject
+    {
+        //Health
+        [Header("Health")]
+        [SerializeField /*,PropertyRange(0, "_healthThresholdMaximum")*/ ,MinValueAttribute(0.0f), MaxValueAttribute(1.0f), Tooltip("The percentage of health where Derek will increment its phase")]
```
[the remark this answers is not in the document]
Comment (Gursaanj) [reply Assets/_BForBoss/_Core/Scripts/Boss/DerekPhaseDataSO.cs:11]: Good catch! will remove

**Review by m4rtint — APPROVED** [review]

**Pull request merged**